Python/Product/Wsl/Debugger/WslLauncher.cs

```diff
@@ -0,0 +1,327 @@
+                FileName = _bashPath.Value ?? throw new InvalidOperationException("WSL is not installed"),
+                Arguments = QuoteSingleArgument(FixPath(sh)),
+                WorkingDirectory = config.WorkingDirectory,
+                UseShellExecute = false
+            };
+
+            var dte = (EnvDTE.DTE)_serviceProvider.GetService(typeof(EnvDTE.DTE));
+
+            using (var proc = Process.Start(psi)) {
+                AttachToProcess(
+                    dte,
+                    proc,
+                    PythonRemoteDebugPortSupplier.PortSupplierGuid,
+                    "tcp://{0}@127.0.0.1:{1}".FormatInvariant(secret, port)
+                );
+            }
+        }
+
+        private static bool AttachToProcess(EnvDTE.DTE dte, Process processOutput, Guid portSupplier, string transportQualifierUri) {
+            var debugger3 = (EnvDTE90.Debugger3)dte.Debugger;
+            var transports = debugger3.Transports;
+            EnvDTE80.Transport transport = null;
+            for (int i = 1; i <= transports.Count; ++i) {
+                var t = transports.Item(i);
+                if (Guid.Parse(t.ID) == portSupplier) {
+                    transport = t;
+                    break;
+                }
+            }
+            if (transport == null) {
+                return false;
+            }
+
+            var processes = debugger3.GetProcesses(transport, transportQualifierUri);
+            if (processes.Count < 1) {
+                return false;
+            }
+
+            var process = processes.Item(1);
+            return AttachToProcess(processOutput, process);
+        }
+
+        private static bool AttachToProcess(Process processOutput, EnvDTE.Process process, Guid[] engines = null) {
+            // Retry the attach itself 3 times before displaying a Retry/Cancel
+            // dialog to the user.
+            var dte = process.DTE;
+            dte.SuppressUI = true;
+            try {
+                try {
+                    if (engines == null) {
+                        process.Attach();
+                    } else {
+                        var process3 = process as EnvDTE90.Process3;
+                        if (process3 == null) {
+                            return false;
+                        }
+                        process3.Attach2(engines.Select(engine => engine.ToString("B")).ToArray());
+                    }
+                    return true;
+                } catch (COMException) {
+                    if (processOutput.WaitForExit(500)) {
+                        // Process exited while we were trying
+                        return false;
+                    }
+                }
+            } finally {
+                dte.SuppressUI = false;
+            }
+
+            // Another attempt, but display UI.
+            process.Attach();
+            return true;
+        }
+
+
+        private static string FixPath(string path) {
```

Comment:
Oh, that's a nice trick. Should be more reliable than what's currently there.

I'll add a TODO - none of this code is shipping yet